gvt (and gb-vendor) importer

michael-go opened the pull request:
<!--
Work-in-progress PRs are welcome as a way to get early feedback - just prefix
the title with [WIP].
-->

### What does this do / why do we need it?

It imports the config (`vendor/manifest`) of https://github.com/FiloSottile/gvt

### What should your reviewer look out for in this PR?

`gvt` and `gb-vendor` use the same filename for their manifests, and have a pretty similar structure, as `gvt` is based on `gb-vendor`.
There already is a PR open for `gb`: https://github.com/golang/dep/pull/818, however it's based on the code prior to the recently merged "Standardize the importers" https://github.com/golang/dep/pull/1100, and @carolynvs suggested I opened a new PR anyway.

### Do you need help or clarification on anything?

Every import in the `gvt` manifest has a `repository` field, which has the URL of the repository. On one project I tested the importer with, it had the `repository` field of all `golang.org/x/crypto` packages point to `https://go.googlesource.com/crypto`.  To make the import more accurate, I tried to init the `importedPackage` struct via `Source: pkg.Repository`, however it fails with: 

```unable to list versions for golang.org/x/crypto(https://go.googlesource.com/crypto): unable to deduce repository and source type for "https://go.googlesource.com/crypto": unable to read metadata: go-import metadata not found```

Thanks!

### Which issue(s) does this PR fix?

fixes https://github.com/golang/dep/issues/679

<!--

fixes #
fixes #

-->

michael-go commented:
I signed it!

sdboyer commented:
hmm, so, re: the `Source` issue, I think we may need to do a little extra work in order to determine if the named source is one of the default possible ones for a given import path, and if so, omit it. that will require a new method on `SourceManager`

carolynvs commented:
So `gb`and `gvt` use the exact same file format, and gvt indicates that it created the file by inserting `"generator": "github.com/FiloSottile/gvt",` into the manifest? If that's right, then we won't need two importers and both can be handled by a single importer.

I don't know enough about either project to say for sure, and it shouldn't impact this PR. Just thinking about whether or not we can close the gb PR when this merges.

carolynvs commented:
RE: the source field, [this is why the original gb PR got hung up](https://github.com/golang/dep/pull/818#discussion_r130129733). @michael-go Let us know if you need have questions or help implementing what Sam is suggesting.

michael-go commented on cmd/dep/gvt_importer.go:
```diff
@@ -0,0 +1,117 @@
+		}
+
+		var contstraintHint = ""
+		if pkg.Branch != "master" {
```
I added this line because it actually affected the behavior. 
for example, I ran it on a `gvt` project with this in the `manifest`:
```json
{                                                                        
  "importpath": "github.com/blang/semver",                                                                                                                 
  "repository": "https://github.com/blang/semver",                     
  "vcs": "",                                                           
  "revision": "aea32c919a18e5ef4537bbd283ff29594b1b0165",              
  "branch": "master"                                                   
}, 
```
by passing the `"master" ` as a `ConstraintHint`, `dep init`  chose a newer version:
```toml
[[projects]]
  name = "github.com/blang/semver"
  packages = ["."]
  revision = "2ee87856327ba09384cabd113bc6b5d174e9ec0f"
  version = "v3.5.1"
```

but by not passing `"master"` as a hint with the check above, `dep init` chose the correct version:
```toml
[[projects]]
  name = "github.com/blang/semver"
  packages = ["."]
  revision = "aea32c919a18e5ef4537bbd283ff29594b1b0165"
  version = "v3.1.0"
````

michael-go commented:
@sdboyer, regarding the `Source`. So if I understand correctly, you suggest a new method in `SourceManger`, something like `IsDefaultSource(ProjectIdentifier) bool`, and this method will be called from `baseImporter.importPackages()`, and if returns true, the `.Source` will be cleared from the `gps.ProjectIdentifier` prior the first call to `InferConstraint()`?
If so, you would like this to be part of this PR?
Thanks!

and @carolynvs, thanks! I'll rebase my branch with the changes above when https://github.com/golang/dep/pull/1145 is merged

sdboyer commented:
yep! that sounds ideal.

On September 12, 2017 3:49:27 AM EDT, michael-go <notifications@github.com> wrote:
>@sdboyer, regarding the `Source`. So if I understand correctly, you
>suggest a new method in `SourceManger`, something like
>`IsDefaultSource(ProjectIdentifier) bool`, and this method will be
>called from `baseImporter.importPackages()`, and if returns true, the
>`.Source` will be cleared from the `gps.ProjectIdentifier` prior the
>first call to `InferConstraint()`?
>If so, you would like this to be part of this PR?
>Thanks!
>
>and @carolynvs, thanks! I'll rebase my branch with the changes above
>when https://github.com/golang/dep/pull/1145 is merged
>
>-- 
>You are receiving this because you were mentioned.
>Reply to this email directly or view it on GitHub:
>https://github.com/golang/dep/pull/1149#issuecomment-328770471

michael-go commented:
@carolynvs I rebased the branch following the merge of #1445.

I still check that the `.Branch` isn't "master" before passing it as a `ConstraintHint`, because otherwise it overrides the imported revision, as in the example I gave above (https://github.com/golang/dep/pull/1149#discussion_r137990754). `dep init` prints:
```
(0)   ✓ select (root)
(1)	? attempt github.com/blang/semver with 1 pkgs; at least 1 versions to try
(1)	    try github.com/blang/semver@v3.1.0
(2)	✗   github.com/blang/semver@v3.1.0 not allowed by constraint master:
(2)	    master from (root)
(1)	    try github.com/blang/semver@v3.5.1
(1)	✓ select github.com/blang/semver@v3.5.1 w/1 pkgs
```

Regarding the `repository` field in `gvt`'s manifest, I'm waiting for the merge of https://github.com/golang/dep/pull/1166 I've opened following the helpful advice of @sdboyer.

Thanks!

carolynvs commented:
I was able to reproduce the issue of the master constraint.

The problem is that imported revision is tagged with v3.1.0, and the base importer is using that version in the lock, while still using a branch constraint in the manifest, which is invalid dep configuration. Solve will detect that v3.1.0 doesn't satisfy the branch constraint, and pick a different revision that satisfies the constraint.

I am still thinking about how to resolve this:

1. The base importer can ignore that the revision is tagged, and instead of setting `version = "v3.1.0"` in the lock, use `branch = "master"` instead.
2. The base importer can ignore the master constraint, and continue setting `version = "v3.1.0"` in the lock. It could then either use an any constraint, or `^3.1.0` in the manifest.

Either of those solutions would preserve the correct locked revision. The first is "truer" to the original imported config, preserving the master branch constraint. However, dep would much prefer to move people onto using semver constraints.

Plus I'm not sure that gvt really meant "stay on master" instead of "this revision happened to be the tip of master when gvt was run". 🤔

michael-go commented:
Thanks @carolynvs. 
I updated the docs with `gvt` and also added the integration test for it, which b.t.w. catches the `"branch": "master"` issue with https://github.com/sdboyer/deptestdos 

regarding `gb` - I'll look later at https://github.com/golang/dep/pull/818 to make sure I cover here all the cases for it

michael-go commented:
OK, so after looking at https://github.com/golang/dep/pull/818, I added a special handling to a case where "branch" is set "HEAD", and updated the docs (and console output) to include `gb`

carolynvs commented on internal/importers/gvt/importer.go:
```diff
@@ -0,0 +1,130 @@
+
+		ip := base.ImportedPackage{
+			Name: pkg.ImportPath,
+			//TODO: temporarly ignore .Repository. see https://github.com/golang/dep/pull/1166
```
I think once you get a chance to incorporate your PR back into this, we can merge.

michael-go replied:
thanks @carolynvs ! will try to add a commit tomorrow

michael-go commented:
@carolynvs I rebased over the new master following the merge of https://github.com/golang/dep/pull/1166 and handled the `Respository` issue. Thanks! :pray:

carolynvs commented on internal/importers/base/importer.go:
```diff
@@ -180,10 +180,18 @@ func (i *Importer) ImportPackages(packages []ImportedPackage, defaultConstraintF
+		source := prj.Source
+		if len(source) > 0 {
+			isDefault, err := i.isDefaultSource(prj.Root, source)
+			if err == nil && isDefault {
```
Let's print a warning, and ignore the imported source. Then they can review the warning and tweak if necessary.

Also add a testcase in `TestBaseImporter_ImportProjects` to verify that `WantWarning` contains a substring from our warning message and that the source was set to `""`. 

e.g.

```
Ignoring imported source https://github.com/example/foo.blorp for github.com/example/foo: <error string>
```

michael-go replied:
Do you mean it for every time we set `Source` to `""` , or just when `isDefaultSource()` returns an error? (and it should actually never return an error, as `DeduceRootProject()` that is called earlier in `loadPackages` calls the same underlying `deduceRootPath()`)

carolynvs replied:
Sorry that wasn't clear! Yes, I was referring to when `isDefaultSource` returns an error. Even though it may never happen, I'd prefer to handle it regardless.

carolynvs commented on internal/importers/base/importer.go:
```diff
@@ -291,3 +299,20 @@ func (i *Importer) convertToConstraint(v gps.Version) gps.Constraint {
 }
+
+func (i *Importer) isDefaultSource(projectRoot gps.ProjectRoot, sourceURL string) (bool, error) {
+	if sourceURL == "https://"+string(projectRoot) {
```
Can you help me understand why this is necessary? I would have thought that `SourceURLsForPath` would cover this as well.

michael-go replied:
mainly for `gopkg.in` imports, as for example importing `gopkg.in/yaml.v2` via `gvt` will appear in the manifest as:
```json
{
	"importpath": "gopkg.in/yaml.v2",
	"repository": "https://gopkg.in/yaml.v2",
	"vcs": "",
	"revision": "f7716cbe52baa25d2e9b0d0da546fcf909fc16b4",
	"branch": "v2"
}
```
but `SourceURLsForPath()` will return `[https://github.com/airbrake/gobrake http://github.com/airbrake/gobrake]`

In addition, this small condition will usually return true and will same save some CPU cycles :smile:

carolynvs replied:
Would you please add a comment explaining that we are essentially checking for gopkg.in URLs?

carolynvs commented on internal/importers/base/importer.go:
```diff
@@ -291,3 +299,20 @@ func (i *Importer) convertToConstraint(v gps.Version) gps.Constraint {
+		return false, err
+	}
+	// The first url in the slice will be the default one (usually https://...)
+	if len(sourceURLs) > 0 && sourceURL == sourceURLs[0].String() {
```
If the first entry in `sourceURLs` doesn't exist, does dep try the other ones? e.g. If the imported source matched the second entry, but we don't import it and left the source blank, would dep automatically try the first, realize it doesn't exist and fall back to that second entry?

I am a bit confused as to why only the first entry is the only "default". I thought they all were defaults that dep would try?

michael-go replied:
regarding you first question, yes, I believe dep will try the next URL in order, if one doesn't exist.

I compare only to the first one following the guidance of @sdboyer - good change I misunderstood something though. My guess for this, is that if a tool choose not the `https:` URL, but the `ssh` one for example, we better respect that and not try fetching via `https` even if it works.

carolynvs replied:
Sounds good to me, thanks!

carolynvs commented on internal/importers/gvt/importer.go:
```diff
@@ -0,0 +1,129 @@
+	gvtConfig gvtManifest
+}
+
+// NewImporter for gvt.
```
So we don't forget, let's include in this comment that it also automatically handles gb as well.

carolynvs commented on internal/importers/gvt/importer_test.go:
```diff
@@ -0,0 +1,245 @@
+				},
+			},
+		},
+		"package with alternate repository": {
```
I believe this is really testing the logic that you added to the base importer, right? Let's move this testcase into the base importer's tests, as it's not specific to gvt.

michael-go replied:
hm... this tests also tests that I use the `Source` field in `gvt` importer, i.e. it tests this line: https://github.com/michael-go/dep/blob/1cfc27fb29c6369b42d99a51e06e680f1539e819/internal/importers/gvt/importer.go#L116

maybe we just need to added another test in the base importer in addition?

carolynvs replied:
Yeah that sounds good, would you please add a test to the base importer as well?

michael-go commented:
Thanks @carolynvs, I handled all your comments, except I didn't know how to write the test for `isDefaultSource()` error handling, as there is no possible code flow to make the warning to be printed. I can mock `isDefaultSource()` for this, but thought that maybe the 2 tests I added in the last commit are enough. :pray:

carolynvs commented:
Yes, that's enough, thanks!
